Add DEA functions and plotting to ADRIA.analysis

Rosejoycrocker opened the pull request:
Adds function to perform output-oriented (default) Data Envelopment Analysis (DEA) given inputs X and output metrics Y. DEA is used to measure the performance of entities (scenarios), where inputs are converted to outputs via some process. Each scenario's "efficiency score" is calculated relative to an "efficiency frontier", a region representing scenarios for which outputs
cannot be further increased by changing inputs (scenario settings). Scenarios on the frontier serve as "benchmarks" or "peers", associated with best practice restoration scenarios. Scenarios with efficiencies not equal to 1 can be improved to be more efficient.

Allows usage with generic input functions, such as functions representing cost as a function of intervention scenario.

Flagging @ConnectedSystems as  we discussed this previously

Working example is:
```julia
using Revise, Infiltrator
using ADRIA
using DataFrames, YAXArrays
using GLMakie, GeoMakie, GraphMakie
using Distributions

dom = ADRIA.load_domain( "domain path", "45")

scens = ADRIA.sample(dom, 2^10)
rs = ADRIA.run_scenarios(dom, scens, "45")

n_scens = size(scens,1)

# Get cost of deploying corals in each scenario, randomised cost example
cost = YAXArray(collect(range(100,stop=1000000, length=n_scens)) .+ rand(Uniform(1000, 2000), n_scens))

# Get mean coral cover and shelter volume for each scenario
s_tac = dropdims(
    mean(ADRIA.metrics.scenario_total_cover(rs); dims=:timesteps); dims=:timesteps
)
s_sv = dropdims(
   mean(ADRIA.metrics.scenario_shelter_volume(rs); dims=:timesteps); dims=:timesteps
    )

# Do output oriented DEA analysis seeking to maximise cover and shelter volume for minimum
# deployment cost.
DEA_scens = ADRIA.economics.data_envelopment_analysis(cost, Array(s_tac), Array(s_sv))
dea_fig = ADRIA.viz.data_envelopment_analysis(rs, DEA_scens)

```

![example_dea_fig](https://github.com/user-attachments/assets/24f5332b-c868-4dec-bc62-35dc1f64afbe)

Zapiano commented on ext/AvizExt/viz/economics.jl:
```diff
@@ -0,0 +1,129 @@
+    Legend(g[1, 2], [frontier, data], [frontier_name, data_name])
+
+    # Plot the scale efficiency (ratio of efficiencies assuming CRS vs. assuming VRS)
+    axb = Axis(g[2, 1]; title="Scale efficiency", ylabel=scale_eff_y_lab, axis_opts...)
```
```suggestion
    ax_b = Axis(g[2, 1]; title="Scale efficiency", ylabel=scale_eff_y_lab, axis_opts...)
```

Zapiano commented on ext/AvizExt/viz/economics.jl:
```diff
@@ -0,0 +1,129 @@
+    )
+
+    # Plot the technical efficiency (inverse VRS efficiencies)
+    axc = Axis(g[3, 1]; title="Technical efficiency", ylabel=tech_eff_y_lab, axis_opts...)
```
```suggestion
    ax_c = Axis(g[3, 1]; title="Technical efficiency", ylabel=tech_eff_y_lab, axis_opts...)
```

Zapiano reviewed (commented):
This PR doesn't add any new test. I think it would be a good idea to add at least some basic tests, at least the "happy case" both for the metrics and for the viz function added.

Rosejoycrocker commented:
> This PR doesn't add any new test. I think it would be a good idea to add at least some basic tests, at least the "happy case" both for the metrics and for the viz function added.

I've added the plotting and calculation functions to analysis.jl tests

Rosejoycrocker commented on src/analysis/economics.jl:
```diff
@@ -0,0 +1,169 @@
+
+"""
+function data_envelopment_analysis(rs::ResultSet, input_function::Function,
+    metrics...; orient::Symbol=:Output, dea_model::Function=DEA.deabigdata
```
This is the input that the DEA function takes, which is a separate package so I can't change it

Rosejoycrocker commented on src/metrics/scenario.jl:
````diff
@@ -279,3 +282,72 @@ function scenario_outcomes(rs::ResultSet, metrics::Vector{<:Metric})::YAXArray
+outcomes = ADRIA.metrics.scenario_outcomes(rs, metrics)
+```
+"""
+function cf_difference_scenario(rs::ResultSet, metric::Function)
````
This was previously added because the economics team use the difference between counterfactuals and intervention scenarios in their DEA analysis. But I think I'll remove for now and add in a later PR to better integrate with your bootstrapping functions (yours may even be able to do this already with a few tweaks)

Rosejoycrocker commented on ext/AvizExt/viz/economics.jl:
```diff
@@ -0,0 +1,129 @@
+    frontier_color = get(opts, :frontier_color, :red)
+    data_color = get(opts, :data_color, :black)
+    frontier_name = get(opts, :frontier_name, "Best practice frontier")
+    data_name = get(opts, :data_name, "Scenario data cloud")
```
Sorry, what do you mean here?

Zapiano replied:
Sorry, I was probably tired and being picky when I wrote this, you can just ignore this comment.

Rosejoycrocker commented on src/analysis/economics.jl:
```diff
@@ -0,0 +1,169 @@
+    )
+end
+function data_envelopment_analysis(
+    X::Vector{Float64}, Y::Matrix{Float64}; orient::Symbol=:Output,
```
All the wrapper functions allow X and Y to be input as YAXArrays, so the user can use a YAXArray as input without having to convert to vector or Array. I can change the inner function to AbstractArray inputs however, although the function will error if the inputs are not Vector or Matrix type because this is what the DEA model takes. Let me know how you'd like me to proceed.

Zapiano replied:
Sorry, about my second comment, I think I was just wrong. 

But about my first comment: if you want to call `data_envelopment_analysis` passing `(rs, Y, input_function)`, `Y` needs to be a YAXArray (so, if the user has an Array it won't work). Same thing with the method that accepts only `X` as argument. But that's a question for you: do you want to give support for versions of those functions with `Array`s instead of `YAXArray`s? If that's not an issue, then maybe you don't need to change anything..?

Rosejoycrocker replied:
Ah ok I see what you mean, thanks! I think I'll allow Y to be an AbstractArray because this could be calculated with any "cost" function, but X should be a YAXAarry because it'll be the result of an ADRIA metric

Rosejoycrocker commented on src/analysis/economics.jl:
```diff
@@ -0,0 +1,169 @@
+    FDH_peers = DEA.peers(result_FDH)
+
+    return DEAResult(
+        result_CRS.eff,
```
I've removed the constructor now as you're right, it's unnecessary

Rosejoycrocker commented:
I think I've addressed your comments @Zapiano, let me know if you have others

Zapiano commented on ext/AvizExt/viz/data_envelopment.jl:
```diff
@@ -0,0 +1,133 @@
+    - `scale_eff_y_lab` : Label for scale efficiency.
+    - `scale_eff_y_lab` : Label for technical efficiency.
+"""
+function ADRIA.viz.data_envelopment_analysis(
+    rs::ResultSet, DEA_output::DEAResult;
+    axis_opts::OPT_TYPE=DEFAULT_OPT_TYPE(), fig_opts::OPT_TYPE=DEFAULT_OPT_TYPE(),
+    opts::OPT_TYPE=DEFAULT_OPT_TYPE()
+)
```
```suggestion
function ADRIA.viz.data_envelopment_analysis(
    rs::ResultSet, 
    DEA_output::DEAResult;
    axis_opts::OPT_TYPE=DEFAULT_OPT_TYPE(), 
    fig_opts::OPT_TYPE=DEFAULT_OPT_TYPE(),
    opts::OPT_TYPE=DEFAULT_OPT_TYPE()
)
```

Zapiano commented on ext/AvizExt/viz/data_envelopment.jl:
```diff
@@ -0,0 +1,133 @@
+using ADRIA.analysis: DEAResult
+
+"""
+    ADRIA.viz.data_envelopment_analysis(rs::ResultSet, DEA_output::DEAResult;axis_opts=Dict(),
```
This docstring doesn't match the function signature (all `opts` params types are wrong) **and** it should be one method per line (this is how we do in the resto of this project at least).

Zapiano commented on ext/AvizExt/viz/data_envelopment.jl:
```diff
@@ -0,0 +1,133 @@
+
+    return f
+end
+function ADRIA.viz.data_envelopment_analysis!(g::Union{GridLayout,GridPosition},
+    rs::ResultSet, DEA_output::DEAResult; axis_opts::OPT_TYPE=DEFAULT_OPT_TYPE(),
+    opts::OPT_TYPE=DEFAULT_OPT_TYPE()
+)
```
```suggestion
function ADRIA.viz.data_envelopment_analysis!(
    g::Union{GridLayout,GridPosition},
    rs::ResultSet, 
    DEA_output::DEAResult; 
    axis_opts::OPT_TYPE=DEFAULT_OPT_TYPE(),
    opts::OPT_TYPE=DEFAULT_OPT_TYPE()
)
```

Zapiano commented on ext/AvizExt/viz/data_envelopment.jl:
```diff
@@ -0,0 +1,133 @@
+        g, DEA_output; opts=opts, axis_opts=axis_opts
+    )
+end
+function ADRIA.viz.data_envelopment_analysis!(g::Union{GridLayout,GridPosition},
+    DEA_output::DEAResult; axis_opts::OPT_TYPE=DEFAULT_OPT_TYPE(),
+    opts::OPT_TYPE=DEFAULT_OPT_TYPE())
```
```suggestion
function ADRIA.viz.data_envelopment_analysis!(
    g::Union{GridLayout,GridPosition},
    DEA_output::DEAResult; axis_opts::OPT_TYPE=DEFAULT_OPT_TYPE(),
    opts::OPT_TYPE=DEFAULT_OPT_TYPE()
)
```

Zapiano commented on ext/AvizExt/viz/data_envelopment.jl:
```diff
@@ -0,0 +1,133 @@
+    scale_efficiency = DEA_output.crs_vals ./ DEA_output.vrs_vals
+
+    # Plot efficiency frontier and data cloud
+    ax_a = Axis(g[1, 1]; xlabel=metrics_x_lab, ylabel=metrics_y_lab, axis_opts...)
```
I think there's a way of making this more interesting, doing the same thing in less lines and extracting part of this code to another function:

If I understand, the reason why you need to save the result of these two following `scatter!` to variables is to render the Legend. You could also render the Legend in a different way:

```
marker_colors = [frontier_color, data_color]
marker_els = [MarkerElement(;color=m_color) for m_color in marker_colors]
Legend(g[1,2], marker_els, [frontier_name, data_name]; framevisible=false)
```

You can even put these three lines inside a separate function and pass `opts` to it so you don't need to extract these colors and names here. 

And if you do that, you don't need to save those `scatter!` results to variables. **Then**, you can extract these lines that you repeat three times to an external function:

```
    # Plot the scale efficiency (ratio of efficiencies assuming CRS vs. assuming VRS)
    ax_b = Axis(g[2, 1]; title="Scale efficiency", ylabel=scale_eff_y_lab, axis_opts...)
    scatter!(ax_b, scale_efficiency; color=data_color)
    scatter!(
        ax_b,
        best_practice_scens,
        scale_efficiency[best_practice_scens];
        color=frontier_color
    )
```

And, again, you can pass `opts` to this function directly and extract the relevant variables inside it so you don't need to extract them here).

Zapiano replied:
Still about the legend: I've done something similar here: https://github.com/open-AIMS/ADRIA.jl/blob/3acd2a72d79fab22c169a2337f65d1bde11f46a8/ext/AvizExt/viz/scenarios.jl#L290

And a small digression, hope you don't mind:

That makes me think we should have some standardized way of doing this the function I'm suggesting to you and the function in this link are essentially the same except that one generates a `Legend` with `LineElement`s and the other generates a `Legend` with `MarkerElement`s - and that's something we could address using multiple dispatch probably.

All that to say: maybe not do what I just said in this PR, but if you just extract this code that generates a legend (the one I suggested in the previous comment) to a function called `render_marker_legend`, for example, in the future we can make this better and have just a `render_legend` that can be used for different types of elements.

Rosejoycrocker replied:
I'll add `render_marker_legend` to data_envelopment.jl` but I'm guessing this will eventually be moved to 'viz.jl' when multiple dispatch is incorporated, so it can be used by both your `scenario.jl` functions and the `data_envelopment.jl` functions?

Zapiano commented on src/analysis/data_envelopment.jl:
```diff
@@ -0,0 +1,143 @@
+    data_envelopment_analysis(rs::ResultSet, input_function::Function, metrics...; orient::Symbol=:Output, dea_model::Function=DEA.deabigdata)::DEAResult
+    data_envelopment_analysis(rs::ResultSet, Y::YAXArray, input_function::Function; orient::Symbol=:Output, dea_model::Function=DEA.deabigdata)::DEAResult
+    data_envelopment_analysis(X::YAXArray, metrics...; orient::Symbol=:Output, dea_model::Function=DEA.deabigdata)::DEAResult
+    data_envelopment_analysis(X::YAXArray, Y::YAXArray; orient::Symbol=:Output, dea_model::Function=DEA.deabigdata)::DEAResult
```
This one doesn't match the function signature.

Rosejoycrocker commented:
I think I've addressed your comments @Zapiano, let me know if there is anything else :)

Rosejoycrocker commented:
> I think I've addressed your comments @Zapiano, let me know if there is anything else :)

Hey @Zapiano, could you give this a final review :)

Zapiano commented:
@Rosejoycrocker apparently the tests are not passing (I checked and they are passing on the main branch):

```
scenario_juvenile_indicator: Test Failed at C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\test_metrics_helper.jl:39
  Expression: all(metric_result.data .<= 1.0)

Stacktrace:
 [1] macro expansion
   @ C:\Users\pribeiro\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:679 [inlined]
 [2] test_metric(metric::ADRIA.metrics.Metric{typeof(ADRIA.metrics._scenario_juvenile_indicator), Tuple{Symbol, Symbol}, String, Bool}, params::Tuple{YAXArray{Float64, 3, Array{Float64, 3}, Tuple{Dim{:timesteps, DimensionalData.Dimensions.Lookups.Sampled{Int64, UnitRange{Int64}, DimensionalData.Dimensions.Lookups.ForwardOrdered, DimensionalData.Dimensions.Lookups.Regular{Int64}, DimensionalData.Dimensions.Lookups.Points, DimensionalData.Dimensions.Lookups.NoMetadata}}, Dim{:species, DimensionalData.Dimensions.Lookups.Sampled{Int64, UnitRange{Int64}, DimensionalData.Dimensions.Lookups.ForwardOrdered, DimensionalData.Dimensions.Lookups.Regular{Int64}, DimensionalData.Dimensions.Lookups.Points, DimensionalData.Dimensions.Lookups.NoMetadata}}, Dim{:locations, DimensionalData.Dimensions.Lookups.Sampled{Int64, UnitRange{Int64}, DimensionalData.Dimensions.Lookups.ForwardOrdered, DimensionalData.Dimensions.Lookups.Regular{Int64}, DimensionalData.Dimensions.Lookups.Points, DimensionalData.Dimensions.Lookups.NoMetadata}}}, Dict{Symbol, Any}}, DataFrame, Vector{Float64}})
   @ Main C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\test_metrics_helper.jl:39
 [3] macro expansion
   @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:58 [inlined]
 [4] macro expansion
   @ C:\Users\pribeiro\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:1704 [inlined]
 [5] macro expansion
   @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:54 [inlined]
 [6] macro expansion
   @ C:\Users\pribeiro\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:1704 [inlined]
 [7] top-level scope
   @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:8
scenario_juvenile_indicator: Test Failed at C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\test_metrics_helper.jl:39
  Expression: all(metric_result.data .<= 1.0)

Stacktrace:
 [1] macro expansion
   @ C:\Users\pribeiro\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:679 [inlined]
 [2] test_metric(metric::ADRIA.metrics.Metric{typeof(ADRIA.metrics._scenario_juvenile_indicator), Tuple{Symbol, Symbol}, String, Bool}, params::Tuple{YAXArray{Float64, 3, Array{Float64, 3}, Tuple{Dim{:timesteps, DimensionalData.Dimensions.Lookups.Sampled{Int64, UnitRange{Int64}, DimensionalData.Dimensions.Lookups.ForwardOrdered, DimensionalData.Dimensions.Lookups.Regular{Int64}, DimensionalData.Dimensions.Lookups.Points, DimensionalData.Dimensions.Lookups.NoMetadata}}, Dim{:species, DimensionalData.Dimensions.Lookups.Sampled{Int64, UnitRange{Int64}, DimensionalData.Dimensions.Lookups.ForwardOrdered, DimensionalData.Dimensions.Lookups.Regular{Int64}, DimensionalData.Dimensions.Lookups.Points, DimensionalData.Dimensions.Lookups.NoMetadata}}, Dim{:locations, DimensionalData.Dimensions.Lookups.Sampled{Int64, UnitRange{Int64}, DimensionalData.Dimensions.Lookups.ForwardOrdered, DimensionalData.Dimensions.Lookups.Regular{Int64}, DimensionalData.Dimensions.Lookups.Points, DimensionalData.Dimensions.Lookups.NoMetadata}}}, Dict{Symbol, Any}}, DataFrame, Vector{Float64}})
   @ Main C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\test_metrics_helper.jl:39
 [3] macro expansion
   @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:58 [inlined]
 [4] macro expansion
   @ C:\Users\pribeiro\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:1704 [inlined]
 [5] macro expansion
   @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:54 [inlined]
 [6] macro expansion
   @ C:\Users\pribeiro\.julia\juliaup\julia-1.11.2+0.x64.w64.mingw32\share\julia\stdlib\v1.11\Test\src\Test.jl:1704 [inlined]
 [7] top-level scope
   @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:8
Test Summary:                 | Pass  Fail  Total   Time
scenario.jl                   |  128     2    130  33.5s
  scenario_total_cover        |   25           25   7.3s
  scenario_relative_cover     |    6            6   0.9s
  scenario_relative_juveniles |   24           24   4.8s
  scenario_absolute_juveniles |   20           20   1.4s
  scenario_juvenile_indicator |   22     2     24   4.6s
  scenario_asv                |   20           20   5.3s
  scenario_rsv                |    6            6   0.8s
  scenario_evenness           |    5            5   3.8s
ERROR: LoadError: Some tests did not pass: 128 passed, 2 failed, 0 errored, 0 broken.
in expression starting at C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\metrics\scenario.jl:7
in expression starting at C:\Users\pribeiro\AIMS\Code\ADRIA.jl\test\runtests.jl:23
ERROR: Package ADRIA errored during testing
```

Rosejoycrocker commented:
> @Rosejoycrocker apparently the tests are not passing (I checked and they are passing on the main branch):
> 

Sorry @Zapiano, I forgot to push my rebase with main, tests are passing on my end now

Rosejoycrocker commented:
Hey @Zapiano could you review this again? I just rebased with main and tests are passing

Zapiano commented:
> Hey @Zapiano could you review this again? I just rebased with main and tests are passing

Hey @Rosejoycrocker it looks like codecov check is failing because the test coverage is decreasing. It looks like a few of your `data_envelopment_analysis` functions and `viz.taxonomy` are not being tested.

Rosejoycrocker commented:
> > Hey @Zapiano could you review this again? I just rebased with main and tests are passing
> 
> Hey @Rosejoycrocker it looks like codecov check is failing because the test coverage is decreasing. It looks like a few of your `data_envelopment_analysis` functions and `viz.taxonomy` are not being tested.

Hmm I've added tests for them here https://github.com/open-AIMS/ADRIA.jl/blob/382c4b90c3044fcea78eb17cdb5a60ae3764ea37/test/analysis.jl#L359-360 but I'll see what else I can add

ConnectedSystems commented:
Looking at the code coverage report I'm unsure there is anything to do with `viz.taxonomy`. I think it's having trouble differentiating with the addition of `data_envelopment_analysis()`.

I suggest ignoring that particular "error" for now, as I don't think it impacts anything given the reported change is 0%.

Regardless, the set of changes in this PR does not affect the taxonomy visualization.

@arlowhite can advise more perhaps.
